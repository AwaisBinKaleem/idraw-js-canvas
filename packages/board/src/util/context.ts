import { TypeContext } from '@idraw/types';

type Options = {
  width: number;
  height: number;
  devicePixelRatio: number;
}

type Config = {
  scale?: number;
  scrollX?: number;
  scrollY?: number;
}

type PrivateConfig = {
  scale: number;
  scrollX: number;
  scrollY: number;
}

class Context implements TypeContext {
  private _opts: Options;
  private _ctx: CanvasRenderingContext2D;
  private _conf: PrivateConfig; 

  // private _scale: number;
  // private _scrollX: number;
  // private _scrollY: number;

  constructor(ctx: CanvasRenderingContext2D, opts: Options) {
    this._opts = opts;
    this._ctx = ctx;
    this._conf = {
      scale: 1,
      scrollX: 0,
      scrollY: 0,
    }
  }

  setConfig(config: Config) {
    this._conf = {...this._conf, ...config};
  }

  setFillStyle(color: string) {
    this._ctx.fillStyle = color;
  }

  fillRect(x: number, y: number, w: number, h: number) {
    return this._ctx.fillRect(
      this._doSize(x), 
      this._doSize(y),
      this._doSize(w),
      this._doSize(h)
    );
  }

  clearRect(x: number, y: number, w: number, h: number) {
    return this._ctx.clearRect(
      this._doSize(x), 
      this._doSize(y),
      this._doSize(w),
      this._doSize(h)
    );
  }

  beginPath() {
    return this._ctx.beginPath();
  }

  closePath() {
    return this._ctx.closePath();
  }

  lineTo(x: number, y: number) {
    return this._ctx.lineTo(this._doSize(x), this._doSize(y));
  }

  setLineWidth(w: number) {
    this._ctx.lineWidth = w;
  }

  isPointInPath(x: number, y: number) {
    return this._ctx.isPointInPath(this._doX(x), this._doY(y));
  }

  setStrokeStyle(color: string) {
    this._ctx.strokeStyle = color;
  }

  stroke() {
    return this._ctx.stroke();
  }

  private _doSize(num: number) {
    return this._opts.devicePixelRatio * num;
  }

  private _doX(x: number) {
    const { scale, scrollX } = this._conf;
    const _x = (x - scrollX * scale) / scale;
    return this._doSize(_x);
  }

  private _doY(y: number) {
    const { scale, scrollY } = this._conf;
    const _y = (y - scrollY * scale) / scale;
    return this._doSize(_y);
  }

}

export default Context;