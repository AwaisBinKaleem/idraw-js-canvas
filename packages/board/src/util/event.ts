import { TypePoint } from '@idraw/types'

export interface TypeBoardEventArgMap {
  'point': TypePoint;
  'move': TypePoint;
  'moveStart': TypePoint;
  'moveEnd': TypePoint;
  'scale': number;
  'scrollX': number;
  'scrollY': number;
}

export interface TypeBoardEvent {
  on<T extends keyof TypeBoardEventArgMap >(key: T, callback: (p: TypeBoardEventArgMap[T]) => any): void
  off<T extends keyof TypeBoardEventArgMap >(key: T, callback: (p: TypeBoardEventArgMap[T]) => any): void
  trigger<T extends keyof TypeBoardEventArgMap >(key: T, p: TypeBoardEventArgMap[T]): void
}


export class BoardEvent implements TypeBoardEvent {

  private _listeners: Map<string, Function[]>;

  constructor() {
    this._listeners = new Map();
  }

  on<T extends keyof TypeBoardEventArgMap >(eventKey: T, callback: (p: TypeBoardEventArgMap[T]) => any) {
    if (this._listeners.has(eventKey)) {
      const callbacks = this._listeners.get(eventKey);
      callbacks?.push(callback);
      this._listeners.set(eventKey, callbacks || [])
    } else {
      this._listeners.set(eventKey, [callback]);
    }
  }
  
  off<T extends keyof TypeBoardEventArgMap >(eventKey: T, callback: (p: TypeBoardEventArgMap[T]) => any) {
    if (this._listeners.has(eventKey)) {
      const callbacks = this._listeners.get(eventKey);
      if (Array.isArray(callbacks)) {
        for (let i = 0; i < callbacks?.length; i++) {
          if (callbacks[i] === callback) {
            callbacks.splice(i, 1);
            break;
          }
        }
      }
      this._listeners.set(eventKey, callbacks || [])
    }
  }

  trigger<T extends keyof TypeBoardEventArgMap >(eventKey: T, arg: TypeBoardEventArgMap[T]) {
    let callbacks = this._listeners.get(eventKey);
    if (Array.isArray(callbacks)) {
      callbacks.forEach((cb) => {
        cb(arg);
      });
      return true;
    } else {
      return false;
    }
  }

  has(name: string) {
    if (this._listeners.has(name)) {
      const list: Function[] | undefined = this._listeners.get(name);
      if (Array.isArray(list) && list.length > 0) {
        return true;
      }
    }
    return false;
  }

}