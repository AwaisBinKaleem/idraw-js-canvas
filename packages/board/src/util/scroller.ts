import {
  TypeScreenPosition
} from '@idraw/types';

type TypeOptions = {
  width: number,
  height: number,
  devicePixelRatio: number
};


export class Scroller {

  private _displayCtx: CanvasRenderingContext2D;
  private _opts: TypeOptions;

  constructor(
    ctx: CanvasRenderingContext2D,
    opts: TypeOptions
  ) {
    this._displayCtx = ctx;
    this._opts = opts;
  }

  draw(position: TypeScreenPosition) {
    const { width, height } = this._opts;
    const wrapper = this._calc(position);
    // TODO
    if (this._displayCtx) {
      console.log('scroller-wrapper ===', wrapper);
    }

    const ctx = this._displayCtx;
    ctx.globalAlpha = 0.4;
    ctx.fillStyle = wrapper.color;
    if (wrapper.xSize > 0) {
      // x-line
      ctx.fillRect(0, this._doSize(height - wrapper.lineSize), this._doSize(width), this._doSize(wrapper.lineSize));

      // x-slider
      drawBox(ctx, {
        x: this._doSize(wrapper.translateX),
        y: this._doSize(height - wrapper.lineSize),
        w: this._doSize(wrapper.xSize),
        h: this._doSize(wrapper.lineSize),
        r: this._doSize(wrapper.lineSize / 2),
        color: wrapper.color,
      });
    }

    if (wrapper.ySize > 0) {
      // y-line
      ctx.fillRect(this._doSize(width - wrapper.lineSize), 0, this._doSize(wrapper.lineSize), this._doSize(height));
      ctx.globalAlpha = 1;

      // y-slider
      drawBox(ctx, {
        x: this._doSize(width - wrapper.lineSize),
        y: this._doSize(wrapper.translateY),
        w: this._doSize(wrapper.lineSize),
        h: this._doSize(wrapper.ySize),
        r: this._doSize(wrapper.lineSize / 2),
        color: wrapper.color,
      });
    }
  

    ctx.globalAlpha = 1;
    
  }

  private _calc(position: TypeScreenPosition) {
    const { width, height } = this._opts;
    const sliderMinSize = 50;
    const lineSize = 16;
    let xSize = 0;
    let ySize = 0;
    if (position.left <= 0 || position.right <= 0) {
      xSize = Math.max(
        sliderMinSize, width - (
          Math.abs(position.left < 0 ? position.left : 0) + Math.abs(position.right < 0 ? position.right : 0)
        )
      );
      if (xSize >= width) xSize = 0;
    }
    if (position.top <= 0 || position.bottom <= 0) {
      ySize = Math.max(
        sliderMinSize, height - (
          Math.abs(position.top < 0 ? position.top : 0) + Math.abs(position.bottom < 0 ? position.bottom : 0)
        )
      );
      if (ySize >= height) ySize = 0;
    }

    let translateX = 0;
    if (xSize > 0) {
      translateX = width * Math.abs(position.left) / (Math.abs(position.left) + Math.abs(position.right));
      translateX = Math.min(Math.max(0, translateX - xSize / 2), width - xSize);
    }

    let translateY = 0;
    if (ySize > 0) {
      translateY = height * Math.abs(position.top) / (Math.abs(position.top) + Math.abs(position.bottom));
      translateY = Math.min(Math.max(0, translateY - ySize / 2), height - ySize);
    }
    const scrollWrapper = {
      lineSize,
      xSize,
      ySize,
      translateY,
      translateX,
      color: '#e0e0e0'
    };
    return scrollWrapper;
  }

  private _doSize(num: number) {
    return num * this._opts.devicePixelRatio;
  }
}


function drawBox(
  ctx: CanvasRenderingContext2D,
  opts: { x: number, y: number, w: number, h: number, r: number, color: string }
): void {

  const { x, y, w, h, color } = opts;
  let r = opts.r;
  r = Math.min(r, w / 2, h / 2);
  if (w < r * 2 || h < r * 2) {
    r = 0;
  };
  ctx.beginPath();
  ctx.moveTo(x + r, y);
  ctx.arcTo(x + w, y, x + w, y + h, r);
  ctx.arcTo(x + w, y + h, x, y + h, r);
  ctx.arcTo(x, y + h, x, y, r);
  ctx.arcTo(x, y, x + w, y, r);
  ctx.closePath();  
  ctx.fillStyle = color;
  ctx.fill(); 
}