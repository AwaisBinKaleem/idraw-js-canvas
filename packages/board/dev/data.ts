import type { Data } from '@idraw/types';
import { createUUID, parseSVGPath } from '@idraw/util';

const data: Data = {
  elements: [
    // {
    //   uuid: createUUID(),
    //   type: 'image',
    //   x: 100,
    //   y: 100,
    //   w: 100,
    //   h: 100,
    //   angle: 30,
    //   detail: {
    //     // src: '/public/images/lena.png?t=002'
    //     src: '@assets/0ee02fb0-ec43-6b9a-9ab4-55be5816ca4a'
    //   },
    //   operations: {
    //     limitRatio: true
    //   }
    // },
    {
      uuid: createUUID(),
      x: 2,
      y: 2,
      w: 100,
      h: 100,
      type: 'circle',
      detail: {
        background: '#f44336'
      }
    },
    {
      uuid: createUUID(),
      type: 'rect',
      x: 50,
      y: 50,
      w: 100,
      h: 100,
      detail: {
        background: '#8bc34a'
      }
    },
    // {
    //   uuid: createUUID(),
    //   type: 'image',
    //   x: 250,
    //   y: 250,
    //   w: 100,
    //   h: 100,
    //   detail: {
    //     // src: '/public/images/github.png?t=003'
    //     src: '@assets/25a0e630-a958-a522-29a4-5dccf04985da'
    //   }
    // },
    // {
    //   uuid: createUUID(),
    //   type: 'group',
    //   x: 250,
    //   y: 250,
    //   w: 100,
    //   h: 100,
    //   detail: {
    //     children: [
    //       {
    //         uuid: createUUID(),
    //         type: 'image',
    //         x: 0,
    //         y: 0,
    //         w: 100,
    //         h: 100,
    //         detail: {
    //           // src: '/public/images/github.png?t=003'
    //           src: '@assets/489c64d2-f642-3855-7bd6-cb40624b62c3'
    //         }
    //       }
    //     ],
    //     // assets: {
    //     //   '@assets/489c64d2-f642-3855-7bd6-cb40624b62c3': {
    //     //     type: 'image',
    //     //     value: '/public/images/github.png?t=003'
    //     //   }
    //     // }
    //   }
    // },
    {
      uuid: createUUID(),
      x: 0,
      y: 300,
      w: 100,
      h: 100,
      type: 'circle',
      detail: {
        background: '#009688'
      }
    },
    {
      uuid: createUUID(),
      x: 300,
      y: 300,
      w: 100,
      h: 100,
      type: 'circle',
      detail: {
        background: '#673ab7'
      }
    },
    {
      uuid: createUUID(),
      x: 300,
      y: 0,
      w: 100,
      h: 100,
      type: 'circle',
      detail: {
        background: '#ffc107'
      }
    },
    {
      uuid: createUUID(),
      x: 150,
      y: 150,
      w: 100,
      h: 100,
      // angle: 30,
      type: 'rect',
      detail: {
        background: '#4caf50',
        clipPath: {
          commands: parseSVGPath(`
          M 10,30
          A 20,20 0,0,1 50,30
          A 20,20 0,0,1 90,30
          Q 90,60 50,90
          Q 10,60 10,30 z`),
          // fill: '#ff0000',
          originX: 10,
          originY: 10,
          originH: 80,
          originW: 80
          // background: '#0000002A'
        }
      }
    },
    {
      uuid: createUUID(),
      x: 100,
      y: 10,
      w: 80,
      h: 80,
      // angle: 30,
      type: 'path',
      detail: {
        commands: parseSVGPath(`
        M 10,30
        A 20,20 0,0,1 50,30
        A 20,20 0,0,1 90,30
        Q 90,60 50,90
        Q 10,60 10,30 z`),
        fill: 'red',
        originX: 10,
        originY: 10,
        originH: 80,
        originW: 80,
        background: '#0000002A'

        // clipPath: {
        //   commands: parseSVGPath(`
        //     M 10,30
        //     A 20,20 0,0,1 50,30
        //     A 20,20 0,0,1 90,30
        //     Q 90,60 50,90
        //     Q 10,60 10,30 z`),
        //   fill: '#ff0000',
        //   originX: 10,
        //   originY: 10,
        //   originH: 80,
        //   originW: 80
        // }
      }
    }
  ]
  // assets: {
  //   '@assets/0ee02fb0-ec43-6b9a-9ab4-55be5816ca4a': {
  //     type: 'image',
  //     value: '/public/images/lena.png?t=003'
  //   },
  //   '@assets/25a0e630-a958-a522-29a4-5dccf04985da': {
  //     type: 'image',
  //     value: '/public/images/github.png?t=002'
  //   }
  // }
};

export default data;
