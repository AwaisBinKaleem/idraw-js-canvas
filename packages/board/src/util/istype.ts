import util from '@idraw/util'

export default util.istype;