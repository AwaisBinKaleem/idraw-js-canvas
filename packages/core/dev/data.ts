import type { Data } from '@idraw/types';
import { createUUID, parseSVGPath } from '@idraw/util';

const data: Data = {
  elements: [
    {
      uuid: createUUID(),
      type: 'image',
      x: 100,
      y: 100,
      w: 100,
      h: 100,
      angle: 30,
      detail: {
        src: '/images/lena.png?t=002'
      },
      operations: {
        limitRatio: true
      }
    },
    {
      uuid: createUUID(),
      x: 2,
      y: 2,
      w: 100,
      h: 100,
      type: 'circle',
      detail: {
        background: '#f44336'
      }
    },
    {
      uuid: createUUID(),
      type: 'rect',
      x: 50,
      y: 50,
      w: 100,
      h: 100,
      detail: {
        background: '#8bc34a'
      }
    },
    {
      uuid: createUUID(),
      type: 'image',
      x: 250,
      y: 250,
      w: 100,
      h: 100,
      detail: {
        src: '/images/github.png?t=003'
      }
    },
    {
      uuid: createUUID(),
      x: 0,
      y: 300,
      w: 100,
      h: 100,
      type: 'circle',
      detail: {
        background: '#009688'
      }
    },
    {
      uuid: createUUID(),
      x: 300,
      y: 300,
      w: 100,
      h: 100,
      type: 'circle',
      detail: {
        background: '#673ab7'
      }
    },
    {
      uuid: createUUID(),
      x: 300,
      y: 0,
      w: 100,
      h: 100,
      type: 'circle',
      detail: {
        background: '#ffc107'
      }
    },
    {
      uuid: createUUID(),
      x: 150,
      y: 150,
      w: 100,
      h: 100,
      angle: 30,
      type: 'rect',
      detail: {
        background: '#4caf50',
        clipPath: {
          commands: parseSVGPath(`
          M 10,30
          A 20,20 0,0,1 50,30
          A 20,20 0,0,1 90,30
          Q 90,60 50,90
          Q 10,60 10,30 z`),
          // fill: '#ff0000',
          originX: 10,
          originY: 10,
          originH: 80,
          originW: 80
          // background: '#0000002A'
        }
      }
    },
    {
      uuid: createUUID(),
      x: 100,
      y: 10,
      w: 80,
      h: 80,
      // angle: 30,
      type: 'path',
      detail: {
        commands: parseSVGPath(`
        M 10,30
        A 20,20 0,0,1 50,30
        A 20,20 0,0,1 90,30
        Q 90,60 50,90
        Q 10,60 10,30 z`),
        fill: 'red',
        originX: 10,
        originY: 10,
        originH: 80,
        originW: 80,
        background: '#0000002A'

        // clipPath: {
        //   commands: parseSVGPath(`
        //     M 10,30
        //     A 20,20 0,0,1 50,30
        //     A 20,20 0,0,1 90,30
        //     Q 90,60 50,90
        //     Q 10,60 10,30 z`),
        //   fill: '#ff0000',
        //   originX: 10,
        //   originY: 10,
        //   originH: 80,
        //   originW: 80
        // }
      }
    }
  ]
};

export default data;
