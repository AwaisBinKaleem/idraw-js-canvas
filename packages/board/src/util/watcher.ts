import { TypePoint } from '@idraw/types';
import { BoardEvent, TypeBoardEventArgMap } from './event';


export class Watcher {

  private _canvas: HTMLCanvasElement;
  private _isMoving: boolean = false;
  // private _onMove?: TypeWatchCallback;
  // private _onMoveStart?: TypeWatchCallback;
  // private _onMoveEnd?: TypeWatchCallback;
  private _event: BoardEvent;

  constructor(canvas: HTMLCanvasElement) {
    this._canvas = canvas;
    this._isMoving = false;
    this._initEvent();
    this._event = new BoardEvent;
  }

  on<T extends keyof TypeBoardEventArgMap >(name: T, callback: (p: TypeBoardEventArgMap[T]) => void) {
    this._event.on(name, callback)
  }

  off<T extends keyof TypeBoardEventArgMap >(name: T, callback: (p: TypeBoardEventArgMap[T]) => void) {
    this._event.off(name, callback)
  }

  _initEvent() {
    const canvas = this._canvas;
    canvas.addEventListener('mousedown', this._listenMoveStart.bind(this));
    canvas.addEventListener('mousemove', this._listenMove.bind(this));
    canvas.addEventListener('mouseup', this._listenMoveEnd.bind(this));

    canvas.addEventListener('touchstart', this._listenMoveStart.bind(this));
    canvas.addEventListener('touchmove', this._listenMove.bind(this));
    canvas.addEventListener('touchend', this._listenMoveEnd.bind(this));

    const mouseupEvent = new MouseEvent('mouseup');
    document.querySelector('body')?.addEventListener('mousemove', (e) => {
      // @ts-ignore
      if (e && e.path && e.path[0] !== canvas) {
        canvas.dispatchEvent(mouseupEvent);
      }
    }, false)
  }

  _listenMoveStart(e: MouseEvent|TouchEvent) {
    e.preventDefault();
    const p = this._getPosition(e);
    if (this._isVaildPoint(p)) {
      if (this._event.has('point')) {
        this._event.trigger('point', p);
      }
      if (this._event.has('moveStart')) {
        this._event.trigger('moveStart', p);
      }
    }
    this._isMoving = true;
  }
  
  _listenMove(e: MouseEvent|TouchEvent) {
    e.preventDefault();
    e.stopPropagation();
    if (this._event.has('move') && this._isMoving === true) {
      const p = this._getPosition(e);
      if (this._isVaildPoint(p)) {
        this._event.trigger('move', p);
      }
    }
  }
  
  _listenMoveEnd(e: MouseEvent|TouchEvent) {
    e.preventDefault();
    if (this._event.has('moveEnd')) {
      const p = this._getPosition(e);
      if (this._isVaildPoint(p)) {
        this._event.trigger('moveEnd', p);
      }
    }
    this._isMoving = false;
  }

  _getPosition(e: MouseEvent|TouchEvent) {
    const canvas = this._canvas;
    let x = 0;
    let y = 0;

    if (e instanceof TouchEvent) {
      const touch: Touch = e.touches[0];
      if (touch) {
        x = touch.clientX;
        y = touch.clientY;
      }
    } else {
      x = e.clientX;
      y = e.clientY;
    }

    const p = {
      x: x - canvas.getBoundingClientRect().left,
      y: y - canvas.getBoundingClientRect().top,
      t: Date.now(),
    };
    return p;
  }

  private _isVaildPoint(p: TypePoint) {
    return ( p.x > 0 && p.y > 0)
  }
  
}
