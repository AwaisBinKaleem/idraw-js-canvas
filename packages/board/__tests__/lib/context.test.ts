import Context from './../../src/lib/context';
import util from '@idraw/util'
import { getData } from './../data';

const { deepClone } = util.data;

describe('@idraw/board: src/lib/context', () => {
  const options = {
    width: 600,
    height: 400,
    contextWidth: 1000,
    contextHeight: 900,
    devicePixelRatio: 2
  }

  test('Context', async () => {  
    const opts = deepClone(options);
    const canvas = document.createElement('canvas');
    canvas.width = opts.contextWidth;
    canvas.height = opts.contextHeight;
    const ctx2d: CanvasRenderingContext2D = canvas.getContext('2d') as CanvasRenderingContext2D;
    
    const ctx = new Context(ctx2d, opts);
    const data = getData();
   
    ctx.clearRect(0, 0, opts.contextWidth, opts.contextHeight);
    ctx.setFillStyle('#ffffff');
    ctx.fillRect(0, 0, opts.contextWidth, opts.contextHeight);
    data.elements.forEach(ele => {
      ctx.setFillStyle(ele.desc.color);
      ctx.fillRect(ele.x, ele.y, ele.w, ele.h);
    }); 
    // @ts-ignore;
    const calls = ctx2d.__getDrawCalls();
    expect(calls).toMatchSnapshot();  
  });

  test('Context.getSize', async () => {
    const opts = deepClone(options);
    const canvas = document.createElement('canvas');
    canvas.width = opts.contextWidth;
    canvas.height = opts.contextHeight;
    const ctx2d: CanvasRenderingContext2D = canvas.getContext('2d') as CanvasRenderingContext2D;
    const ctx = new Context(ctx2d, opts);
    expect(ctx.getSize()).toStrictEqual(opts);
  })

  test('Context.resetSize', async () => { 
    const opts = deepClone(options);
    const canvas = document.createElement('canvas');
    canvas.width = opts.contextWidth;
    canvas.height = opts.contextHeight;
    const ctx2d: CanvasRenderingContext2D = canvas.getContext('2d') as CanvasRenderingContext2D;
    const newOpts = {
      width: 601,
      height: 401,
      contextWidth: 1001,
      contextHeight: 901,
      devicePixelRatio: 3,
    }
    const ctx = new Context(ctx2d, opts);
    ctx.resetSize(newOpts);
    expect(ctx.getSize()).toStrictEqual(newOpts);
  });
  
  test('Context.calcDeviceNum', async () => {  
    const opts = deepClone(options);
    const canvas = document.createElement('canvas');
    canvas.width = opts.contextWidth;
    canvas.height = opts.contextHeight;
    const ctx2d: CanvasRenderingContext2D = canvas.getContext('2d') as CanvasRenderingContext2D;
    const ctx = new Context(ctx2d, opts);
    const num = 100;
    expect(ctx.calcDeviceNum(num)).toStrictEqual(opts.devicePixelRatio * num);
  });

  test('Context.calcScreenNum', async () => {  
    const opts = deepClone(options);
    const canvas = document.createElement('canvas');
    canvas.width = opts.contextWidth;
    canvas.height = opts.contextHeight;
    const ctx2d: CanvasRenderingContext2D = canvas.getContext('2d') as CanvasRenderingContext2D;
    const ctx = new Context(ctx2d, opts);
    const num = 100;
    expect(ctx.calcScreenNum(num)).toStrictEqual(num / opts.devicePixelRatio);
  });

  test('Context.setTransform', async () => {  
    const opts = deepClone(options);
    const canvas = document.createElement('canvas');
    canvas.width = opts.contextWidth;
    canvas.height = opts.contextHeight;
    const ctx2d: CanvasRenderingContext2D = canvas.getContext('2d') as CanvasRenderingContext2D;
    const ctx = new Context(ctx2d, opts);
    const transform = {
      scale: 2,
      scrollX: 100,
      scrollY: -200,
    };
    ctx.setTransform(deepClone(transform))
    expect(ctx.getTransform()).toStrictEqual(deepClone(transform));
  });

  test('Context.getTransform', async () => {  
    const opts = deepClone(options);
    const canvas = document.createElement('canvas');
    canvas.width = opts.contextWidth;
    canvas.height = opts.contextHeight;
    const ctx2d: CanvasRenderingContext2D = canvas.getContext('2d') as CanvasRenderingContext2D;
    const ctx = new Context(ctx2d, opts);
    expect(ctx.getTransform()).toStrictEqual({
      scale: 1,
      scrollX: 0,
      scrollY: 0,
    });
  });


  test('Context.setFillStyle', async () => {
    const opts = deepClone(options);
    const canvas = document.createElement('canvas');
    canvas.width = opts.contextWidth;
    canvas.height = opts.contextHeight;
    const ctx2d: CanvasRenderingContext2D = canvas.getContext('2d') as CanvasRenderingContext2D;
    const ctx = new Context(ctx2d, opts);
    const color = '#f0f0f0';
    ctx.setFillStyle(color);
    ctx.fillRect(0, 0, opts.contextWidth, opts.contextHeight);
    expect(ctx2d.fillStyle).toStrictEqual(color);
  });


  test('Context.fill', async () => {
    const opts = deepClone(options);
    const canvas = document.createElement('canvas');
    canvas.width = opts.contextWidth;
    canvas.height = opts.contextHeight;
    const ctx2d: CanvasRenderingContext2D = canvas.getContext('2d') as CanvasRenderingContext2D;
    const ctx = new Context(ctx2d, opts);
    ctx.fill();
    // @ts-ignore
    const calls = ctx2d.__getDrawCalls();
    expect(calls).toMatchSnapshot();
  })

  test('Context.arc', async () => {
    const opts = deepClone(options);
    const canvas = document.createElement('canvas');
    canvas.width = opts.contextWidth;
    canvas.height = opts.contextHeight;
    const ctx2d: CanvasRenderingContext2D = canvas.getContext('2d') as CanvasRenderingContext2D;
    const ctx = new Context(ctx2d, opts);
    ctx.arc(70, 80, 50, 0, Math.PI * 2, true);
    ctx.fill();
    // @ts-ignore
    const calls = ctx2d.__getDrawCalls();
    // console.log('calls =', JSON.stringify(calls, null, 2));
    expect(calls).toMatchSnapshot();
  })


  test('Context.rect', async () => {
    const opts = deepClone(options);
    const canvas = document.createElement('canvas');
    canvas.width = opts.contextWidth;
    canvas.height = opts.contextHeight;
    const ctx2d: CanvasRenderingContext2D = canvas.getContext('2d') as CanvasRenderingContext2D;
    const ctx = new Context(ctx2d, opts);
    ctx.rect(10, 20, 100, 200);
    ctx.fill();
    // @ts-ignore
    const calls = ctx2d.__getDrawCalls();
    // console.log('calls =', JSON.stringify(calls, null, 2));
    expect(calls).toMatchSnapshot();
  })


  test('Context.fillRect', async () => {
    const opts = deepClone(options);
    const canvas = document.createElement('canvas');
    canvas.width = opts.contextWidth;
    canvas.height = opts.contextHeight;
    const ctx2d: CanvasRenderingContext2D = canvas.getContext('2d') as CanvasRenderingContext2D;
    const ctx = new Context(ctx2d, opts);
    ctx.fillRect(0, 0, opts.contextWidth, opts.contextHeight);
    // @ts-ignore
    const calls = ctx2d.__getDrawCalls();
    // console.log('calls =', JSON.stringify(calls, null, 2));
    expect(calls).toMatchSnapshot();
    // expect(calls).toStrictEqual([
    //   {
    //     type: 'fillRect',
    //     transform: [ 1, 0, 0, 1, 0, 0 ],
    //     props: { x: 0, y: 0, width: 2000, height: 1800 }
    //   }
    // ]);
  })
  
})

