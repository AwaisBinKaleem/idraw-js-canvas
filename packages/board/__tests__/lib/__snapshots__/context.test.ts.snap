// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@idraw/board: src/lib/context Context 1`] = `
Array [
  Object {
    "props": Object {
      "height": 1800,
      "width": 2000,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "height": 1800,
      "width": 2000,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 240,
      "width": 400,
      "x": 20,
      "y": 20,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 240,
      "width": 400,
      "x": 160,
      "y": 160,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 240,
      "width": 400,
      "x": 320,
      "y": 320,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "height": 200,
      "width": 400,
      "x": 780,
      "y": 580,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
]
`;

exports[`@idraw/board: src/lib/context Context.arc 1`] = `
Array [
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "anticlockwise": true,
            "endAngle": 6.283185307179586,
            "radius": 100,
            "startAngle": 0,
            "x": 140,
            "y": 160,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
]
`;

exports[`@idraw/board: src/lib/context Context.fill 1`] = `
Array [
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
]
`;

exports[`@idraw/board: src/lib/context Context.fillRect 1`] = `
Array [
  Object {
    "props": Object {
      "height": 1800,
      "width": 2000,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
]
`;

exports[`@idraw/board: src/lib/context Context.rect 1`] = `
Array [
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "height": 400,
            "width": 200,
            "x": 20,
            "y": 40,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "rect",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
]
`;
